node/src/coordinator.rs

```diff
@@ -736,3 +683,131 @@ impl Drop for ReportCurrentJobGuard {
         *self.currently_running_job_name.lock().unwrap() = "".to_string();
     }
 }
+
+/// returns true if one if one of the following occurs:
+/// - the epoch id changes
+/// - a resharing starts
+/// - the participant set changes
+fn stop_running(
+    new_state: &ContractState,
+    current_running_epoch_id: EpochId,
+    current_participant_set: ParticipantsConfig,
+) -> bool {
+    match new_state {
+        ContractState::Running(new_state) => {
+            if new_state.keyset.epoch_id != current_running_epoch_id {
+                tracing::info!("Epoch id changed.");
+                return true;
+            }
+            if new_state.resharing_state.is_some() {
+                tracing::info!("A resharing started.");
+                return true;
+            }
+            if new_state.participants != current_participant_set {
+                tracing::info!("Participant details changed.");
+                return true;
+            }
+            false
+        }
+        _ => {
+            tracing::info!("No longer in Running state.");
+            true
+        }
+    }
+}
+
+fn make_running_stop_fn(
+    current_running_epoch_id: EpochId,
+    current_participant_set: ParticipantsConfig,
+) -> StopFn {
+    Box::new(move |new_state| {
+        stop_running(
+            new_state,
+            current_running_epoch_id,
+            current_participant_set.clone(),
+        )
+    })
+}
+
+/// returns true if one of the following occurs:
+///     - epoch id changed
+///     - resharing concludes
+///     - key event receiver closes the channel.
```

Comment:
Right and naturally we don't allow participant updates during resharing. Makes sense.

Reply:
Yeah, I was thinking of adding a sanity check here. If we ever encounter the case where the participant set changes during a resharing, then this is a serious bug and means the contract has an issue.

I didn't add the check, because it's unclear how we would want to recover from that error.

Reply:
Right yeah but since resharing can be cancelled it makes sense for the resharing operation to fail first and recovery to happen after.